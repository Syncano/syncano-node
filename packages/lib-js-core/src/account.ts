import { ResponseError } from './errors'
import QueryBuilder from './query-builder'

export interface AccountData {
  id: number
  email: string
  first_name: string
  last_name: string
  is_active: boolean
  has_password: boolean
  metadata: object
}

export interface LoginData extends AccountData {
  account_key: string
}

/**
 * Syncano account query builder
 * @property {Function}
 */
export class Account extends QueryBuilder {
  /**
   * Get details of Syncano account
   */
  public get (authKey: string): Promise<AccountData> {
    const fetch = this.nonInstanceFetch.bind(this)
    return new Promise((resolve, reject) => {
      const headers = {
        'X-API-KEY': authKey
      }

      fetch(this.url(), {}, headers)
        .then((res: AccountData)  => resolve(res))
        .catch((err: ResponseError) => reject(err))
    })
  }

  public login (
    {email, password}: {email: string, password: string}
  ): Promise<LoginData> {
    const fetch = this.nonInstanceFetch.bind(this)
    return new Promise((resolve, reject) => {
      const options = {
        body: JSON.stringify({ email, password }),
        method: 'POST'
      }
      fetch(`${this.url()}auth/`, options)
        .then((res: LoginData) => resolve(res))
        .catch((err: ResponseError) => reject(err))
    })
  }

  public register (
    {email, password}: {email: string, password: string}
  ): Promise<LoginData> {
    const fetch = this.nonInstanceFetch.bind(this)
    return new Promise((resolve, reject) => {
      const options = {
        body: JSON.stringify({ email, password }),
        method: 'POST'
      }

      fetch(`${this.url()}register/`, options)
        .then((res: LoginData) => resolve(res))
        .catch((err: ResponseError) => reject(err))
    })
  }

  private url () {
    return `${this.getSyncanoURL()}/account/`
  }
}

export default Account
