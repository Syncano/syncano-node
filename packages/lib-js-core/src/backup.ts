import { NotFoundError } from './errors'
import QueryBuilder from './query-builder'
import { BackupObject } from './types'

/**
 * Syncano Backups
 *
 * @property {Function}
 */
class Backup extends QueryBuilder {
  /**
   * Backups list
   */
  public async list (): Promise<BackupObject[]> {
    return this.getBackups(this.url())
  }

  /**
   * Create new backup
   */
  public async create (): Promise<BackupObject> {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'POST'
    }
    const headers = {
      'X-API-KEY': this.instance.accountKey
    }

    return fetch(this.url(), options, headers)
  }

  /**
   * Delete backup
   */
  public async delete (id?: number | number[]): Promise<any> {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'DELETE'
    }
    const headers = {
      'X-API-KEY': this.instance.accountKey
    }

    if (id === undefined) {
      const items = await this.list()
      const ids = items.map((item: BackupObject) => item.id)

      return this.delete(ids)
    }

    if (Array.isArray(id)) {
      return Promise.all(
        id.map((item) =>
          fetch(`${this.url()}${item}/`, options, headers).then(() => true).catch(() => false)
        )
      )
    }

    await fetch(`${this.url()}${id}/`, options, headers)

    return undefined
  }

  /**
   * Get backup
   *
   * @deprecated User `find`
   */
  public async get (id: number): Promise<BackupObject | undefined> {
    return this.find(id)
  }

  /**
   *  Get backup by id
   *
   * @example
   * const item = await backup.find(10)
   */
  public async find(id: number): Promise<BackupObject | undefined> {
    try {
      return this.findOrFail(id)
    } catch (err) {
      return undefined
    }
  }

  /**
   *  Get backup by id or throw error
   *
   * @example
   * const item = await backup.findOrFail(10)
   */
  public async findOrFail(id: number): Promise<BackupObject> {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'GET'
    }
    const headers = {
      'X-API-KEY': this.instance.accountKey
    }

    try {
      return fetch(`${this.url()}${id}/`, options, headers)
    } catch (err) {
      throw new NotFoundError(`Backup with id: "${id}" was not found.`)
    }
  }

  /**
   *  Get list of backups by id
   *
   * @example
   * const items = await backup.findMany([10, 11])
   */
  public async findMany(ids: number[]): Promise<BackupObject[]> {
    const groups = await this.list()

    return groups.filter((item) => ids.includes(item.id))
  }

  /**
   * Get last backup
   */
  public async last () {
    const backups = await this.list()

    return backups.pop()
  }

  protected url () {
    const {instanceName} = this.instance

    return `${this.getInstanceURL(instanceName)}/backups/full/`
  }

  protected async getBackups (url: string) {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'GET'
    }
    const headers = {
      'X-API-KEY': this.instance.accountKey
    }
    const res = await fetch(url, options, headers)
    let backups = res.objects

    if (res.next) {
      const nextBackups = await this.getBackups(`${this.baseUrl}${res.next}`)

      backups = backups.concat(nextBackups.objects)
    }

    return backups
  }
}

export default Backup
