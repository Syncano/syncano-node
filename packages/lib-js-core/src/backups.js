import QueryBuilder from './query-builder'

/**
 * Syncano Backups
 *
 * @property {Function}
 */
class Backups extends QueryBuilder {
  url () {
    const {instanceName} = this.instance

    return `${this._getInstanceURL(instanceName)}/backups/full/`
  }

  /**
   * Backups list helper
   *
   * @param {string} url
   *
   * @return {Array}
   */
  async _getBackups (url) {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'GET',
    }
    const headers = {
      'x-api-key': this.instance.accountKey
    }
    const res = await fetch(url, options, headers)
    let backups = res.objects

    if(res.next) {
      backups = backups.concat(await this._getGroups(`${this.baseUrl}${res.next}`))
    }

    return backups
  }

  /**
   * Backups list
   *
   * @return {Promise}
   */
  async list () {
    return  this._getBackups(this.url())
  }

  /**
   * Create new backup
   *
   * @return {Promise}
   */
  async create () {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'POST',
    }
    const headers = {
      'x-api-key': this.instance.accountKey
    }

    return fetch(this.url(), options, headers)
  }

  /**
   * Delete backup
   *
   * @param {number} id - Backup id
   *
   * @return {Promise}
   */
  async delete (id) {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'DELETE',
    }
    const headers = {
      'x-api-key': this.instance.accountKey
    }

    return fetch(`${this.url()}${id}/`, options, headers)
  }

  /**
   * Delete all backups
   *
   * @return {Promise}
   */
  async deleteAll () {
    const backups = await this.list()
    backups.forEach(({id}) => this.delete(id))
  }

  /**
   * Get backup
   *
   * @param {number} id - Backup id
   *
   * @return {Promise}
   */
  async get (id) {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'GET',
    }
    const headers = {
      'x-api-key': this.instance.accountKey
    }

    return fetch(`${this.url()}${id}/`, options, headers)
  }

  /**
   * Get last backup
   *
   * @return {Promise}
   */
  async last () {
    const backups = await this.list()
    return backups.pop()
  }
}

export default Backups
