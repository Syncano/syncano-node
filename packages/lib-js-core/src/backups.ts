import QueryBuilder from './query-builder'
import { BackupObject } from './types'

/**
 * Syncano Backups
 *
 * @property {Function}
 */
class Backups extends QueryBuilder {
  /**
   * Backups list
   */
  public async list (): Promise<BackupObject[]> {
    return this.getBackups(this.url())
  }

  /**
   * Create new backup
   */
  public async create (): Promise<BackupObject> {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'POST'
    }
    const headers = {
      'X-API-KEY': this.instance.accountKey
    }

    return fetch(this.url(), options, headers)
  }

  /**
   * Delete backup
   */
  public async delete (id?: number | number[]): Promise<any> {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'DELETE'
    }
    const headers = {
      'X-API-KEY': this.instance.accountKey
    }

    if (id === undefined) {
      const items = await this.list()
      const ids = items.map((item: BackupObject) => item.id)

      return this.delete(ids)
    }

    if (Array.isArray(id)) {
      return Promise.all(
        id.map((item) =>
          fetch(`${this.url()}${item}/`, options, headers)
        )
      )
    }

    return fetch(`${this.url()}${id}/`, options, headers)
  }

  /**
   * Delete all backups
   *
   * @deprecated Use `delete` method without arguments
   */
  public async deleteAll () {
    return this.delete()
  }

  /**
   * Get backup
   */
  public async get (id: number): Promise<BackupObject> {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'GET'
    }
    const headers = {
      'X-API-KEY': this.instance.accountKey
    }

    return fetch(`${this.url()}${id}/`, options, headers)
  }

  /**
   * Get last backup
   */
  public async last () {
    const backups = await this.list()

    return backups.pop()
  }

  protected url () {
    const {instanceName} = this.instance

    return `${this.getInstanceURL(instanceName)}/backups/full/`
  }

  /**
   * Backups list helper
   */
  protected async getBackups (url: string) {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'GET'
    }
    const headers = {
      'X-API-KEY': this.instance.accountKey
    }
    const res = await fetch(url, options, headers)
    let backups = res.objects

    if (res.next) {
      const nextBackups = await this.getBackups(`${this.baseUrl}${res.next}`)

      backups = backups.concat(nextBackups.objects)
    }

    return backups
  }
}

export default Backups
