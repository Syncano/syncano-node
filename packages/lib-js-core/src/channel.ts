import QueryBuilder from './query-builder'

export interface ChannelResponse<T> {
  id: number,
  room: string,
  created_at: string,
  action: string,
  author: {
    admin: number
  },
  metadata: {
    type: string
  },
  payload: T,
  links: {
    self: string
    [x: string]: string
  }
}

export class Channel extends QueryBuilder {
  /**
   * Publish any payload to channel
   *
   * @param channelName A string value used as channel name. Channel types:
   *                    - Public channels: `messages`
   *                    - User private channel: `messages.${username}`
   *                    - Channel room: `messages.${room}`
   * @param payload Value published to channel
   */
  public publish<T> (channelName: string, payload: T): Promise<ChannelResponse<T>> {
    const fetch = this.fetch.bind(this)

    return new Promise((resolve, reject) => {
      const options = {
        body: JSON.stringify({room: channelName, payload}),
        method: 'POST'
      }
      fetch(this.url(), options)
        .then(resolve)
        .catch(reject)
    })
  }

  private url () {
    const {instanceName} = this.instance

    return `${this.getInstanceURL(instanceName)}/channels/default/publish/`
  }
}

export default Channel
