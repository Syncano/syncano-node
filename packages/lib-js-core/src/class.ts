import {QueryBuilder} from './query-builder'
import {SchemaObject, SyncanoClass} from './types'

export class Class extends QueryBuilder {
  public url (className?: string, apiVersion?: string) {
    const {instanceName} = this.instance
    const baseUrl = `${this.getInstanceURL(instanceName, apiVersion)}/classes/`

    return className ? `${baseUrl}${className}/` : baseUrl
  }

  /**
   * Create Syncano Class
   *
   * @param params.name Data Class name - max 64 characters.
   * @param params.description Data Class description
   * @param params.metadata Additional JSON metadata associated with the Class
   * @param params.schema Data Class schema. It defines the Data Objects that
   *                      will be created within the Data Class. `type` (i.e. string)
   *                      and `name` are required fields. The very basic schema would
   *                      look like this: [{"type": "string", "name": "parameter_name"}]
   */
  public create (params: {
    name: string
    description?: string
    schema?: SchemaObject[]
    metadata?: object
  }): Promise<SyncanoClass> {
    const fetch = this.fetch.bind(this)

    return fetch(this.url(), {
      body: JSON.stringify(params),
      method: 'POST'
    })
  }

  /**
   * Delete Syncano Class
   */
  public delete (className: string): Promise<void> {
    const fetch = this.fetch.bind(this)

    return fetch(this.url(className), {
      method: 'DELETE'
    })
  }

  /**
   * Get single Syncano Class details
   */
  public async get (className: string): Promise<SyncanoClass> {
    const fetch = this.fetch.bind(this)
    const resp = await fetch(this.url(className, 'v3'), {
      method: 'GET'
    })

    return resp
  }

  /**
   * List Syncano Classes
   */
  public async list (): Promise<SyncanoClass[]> {
    const fetch = this.fetch.bind(this)
    const resp = await fetch(this.url(undefined, 'v3'), {})

    return resp.objects
  }
}

export default Class
