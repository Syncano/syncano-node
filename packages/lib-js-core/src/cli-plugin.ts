import {writeFileSync} from 'fs'

export default class {
  public context: any

  constructor (context: any) {
    this.context = context
  }
  public async run ([out]: [string]) {
    const sockets = await this.context.Socket.list()
    const classes = Object.assign({}, ...sockets.map((item: any) => item.spec.classes))

    writeFileSync('./syncano.d.ts', this.instanceInterface(classes))
    writeFileSync('./tsconfig.syncano.json', this.syncanoTsConfig)
  }

  private instanceInterface(classes: Array<{
    [className: string]: Array<{
      name: string
      type: string
    }>
  }>) {
    return `import {DataClass} from '@syncano/core'

declare module '@syncano/core/server' {
  export interface InstanceDataSchema {
    ${Object.keys(classes).map((className) => `${className}: DataClass<{
      ${classes[className].map((item: any) => `${item.name}: ${item.type}`).join('\n      ')}
    }>`).join('\n    ')}
  }
}`
  }

  get syncanoTsConfig() {
    return `{
  "files": [
    "./syncano.d.ts"
  ],
  "include": [
    "syncano/**/*"
  ]
}
`
  }
}
