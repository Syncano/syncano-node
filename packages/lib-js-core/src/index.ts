import * as FormData from 'form-data'
import {default as Server} from './server'
export {
  ClassObject,
  User,
  UserGroup,
  BackupObject,
  SyncanoResponse,
  AccountOwner,
  ChannelResponse,
  ClassObjectLinks,
  Hosting,
  HostingFile,
  Instance,
  InstanceMetadata,
  InstanceOwner,
  LoginData,
  SchemaObject,
  Settings,
  Socket,
  SyncanoClass,
  Trace
} from './types'
import {BackupClass} from './backup'
import {DataClass} from './data'
export {Server as Core}
export {DataClass, BackupClass}
export {Context} from './context'
export default Server

module.exports = Server
module.exports.default = Server
module.exports.BackupClass = BackupClass
module.exports.DataClass = DataClass
module.exports.Core = Server
module.exports.FormData = FormData
