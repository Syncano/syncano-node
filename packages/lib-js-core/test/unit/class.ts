import * as chai from 'chai'
import * as chaiAsPromised from 'chai-as-promised'
import * as nock from 'nock'
import Server from '../../src'
import {Class} from '../../src/class'

chai.use(chaiAsPromised)
chai.should()

describe('Class', () => {
  const instanceName = 'testInstance'
  let api: nock.Scope
  // tslint:disable-next-line:variable-name
  let _class: Class

  beforeEach(() => {
    const server = new Server({
      token: 'testKey',
      instanceName
    })
    _class = server._class
    api = nock(`https://${process.env.SYNCANO_HOST || 'api.syncano.io'}`)
  })

  describe('#create', () => {
    it('should create syncano class', () => {
      api
        .post(`/v2/instances/${instanceName}/classes/`, {
          name: 'posts'
        })
        .reply(200)

      return _class.create({name: 'posts'}).should.be.fulfilled
    })
  })

  describe('#delete', () => {
    it('should delete syncano class', () => {
      api.delete(`/v2/instances/${instanceName}/classes/posts/`).reply(200)

      return _class.delete('posts').should.be.fulfilled
    })
  })

  describe('#get', () => {
    it('should return single class details', () => {
      api.get(`/v3/instances/${instanceName}/classes/posts/`).reply(200, {name: 'posts'})

      return expect(_class.get('posts')).resolves.toMatchSnapshot()
    })
  })

  describe('#list', () => {
    it('should return list of syncano classes', () => {
      api
        .get(`/v3/instances/${instanceName}/classes/`)
        .reply(200, {objects: [{name: 'posts'}, {name: 'comments'}]})

      return expect(_class.list()).resolves.toMatchSnapshot()
    })
  })
})
