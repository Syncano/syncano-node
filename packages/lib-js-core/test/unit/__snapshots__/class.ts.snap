// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Class #get should return single class details 1`] = `
Object {
  "name": "posts",
}
`;

exports[`Class #list should return list of syncano classes 1`] = `
Array [
  Object {
    "name": "posts",
  },
  Object {
    "name": "comments",
  },
]
`;
